arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CyclicBarrier.kt

```diff
@@ -23,28 +24,64 @@ public class CyclicBarrier(public val capacity: Int) {
       "Cyclic barrier must be constructed with positive non-zero capacity $capacity but was $capacity > 0"
     }
   }
-  
-  private data class State(val awaiting: Int, val epoch: Long, val unblock: CompletableDeferred<Unit>)
-  
-  private val state: AtomicRef<State> = AtomicRef(State(capacity, 0, CompletableDeferred()))
-  
+
+  private data class State(
+    /** The number of parties that are still required to join before the barrier will open and reset. **/
+    val awaiting: Int,
+    val epoch: Long,
+    val reset: Boolean,
+    val unblock: CompletableDeferred<Unit>,
+    val resetComplete: CompletableDeferred<Unit>
+  )
+
+  private val state: AtomicRef<State> =
+    AtomicRef(State(capacity, 0, false, CompletableDeferred(), CompletableDeferred()))
+
+  /** The number of parties currently waiting. **/
+  public val numberWaiting: Int
+    get() = capacity - state.get().awaiting
+
+  /**
+   * When called, all waiting coroutines will be cancelled with [CancellationException].
+   * When all threads have been cancelled the barrier will cycle.
+   */
+  public suspend fun reset() {
+    state.updateAndGet { s -> s.copy(reset = true) }
+      .resetComplete.await()
+
+    compareAndCycleState(state.get())
+  }
+
+  private fun compareAndCycleState(original: State) = state
+    .compareAndSet(
+      original,
+      State(capacity, original.epoch + 1, false, CompletableDeferred(), CompletableDeferred())
+    )
+
   /**
    * When [await] is called the function will suspend until the required number of coroutines have reached the barrier.
    * Once the [capacity] of the barrier has been reached, the coroutine will be released and continue execution.
    */
   public suspend fun await() {
     state.loop { original ->
-      val (awaiting, epoch, unblock) = original
+      val (awaiting, epoch, reset, unblock, resetComplete) = original
       val awaitingNow = awaiting - 1
-      if (awaitingNow == 0 && state.compareAndSet(original, State(capacity, epoch + 1, CompletableDeferred()))) {
-        unblock.complete(Unit)
-        return
-      } else if (state.compareAndSet(original, State(awaitingNow, epoch, unblock))) {
-        return try {
-          unblock.await()
-        } catch (cancelled: CancellationException) {
-          state.update { s -> if (s.epoch == epoch) s.copy(awaiting = s.awaiting + 1) else s }
-          throw cancelled
+      when {
+        reset -> {
```

Comment:
I think that the `resetComplete` field on `State` could be removed with a call to `unblock.cancel()` but this doesn't seem to work the way I expect - other threads awaiting are not cancelled when I do this.

```
                reset && awaiting == capacity && compareAndCycleState(original) -> {
                    val cancelException = CancellationException("CyclicBarrier was reset.")
                    unblock.cancel(cancelException)
                    throw cancelException
                }
```

Reply:
Resolved with [89457f5](https://github.com/arrow-kt/arrow/pull/3055/commits/89457f5bd5c34ad21476fa024c4e95b0a8631475).